Add qa script for the artifact

karim20230 opened the pull request:
This script verifies the presence of specific artifact files and folders at a given URL. Due to inconsistencies in artifact folder and file structures, this script currently focuses on a subset of artifacts.  Future work will expand its coverage to all artifacts.  This script was developed in collaboration with Kai.

example run the script:
bash ./qa-script.sh https://ghaf-jenkins-controller-release.northeurope.cloudapp.azure.com/artifacts/ghaf-release-pipeline/build_2-commit_40e5ba6995b628057f9e4556c6295e03aba97fd8/aarch64-linux.nvidia-jetson-orin-agx-debug/

Feedback and suggestions are welcome.

joinemm commented on qa-script.sh:
```diff
@@ -0,0 +1,121 @@
+declare -A expected_files=(
+    ["nix-support/hydra-build-products"]=file
+    ["nix-support/system"]=file
+    ["scs/http_cache.sqlite"]=file
```
probably don't need to check for this http cache

karim20230 replied:
Sure, removed

karim20230 commented:
also fixed the dynamic pattern issue

ktusawrk reviewed (changes requested):
**No need to check these files, remove:**

- ["nix-support/hydra-build-products"]=file 
- ["nix-support/system"]=file
- ["esp.offset"]=file 
- ["esp.size"]=file 
- ["root.offset"]=file
- ["root.size"]=file

**Boot and turnoff test result directories have changed, these are the new directories, modify:**

- ["test-results/Robot-Framework/test-suites/relayboot/log.html"]=file
- ["test-results/Robot-Framework/test-suites/relayboot/output.xml"]=file
- ["test-results/Robot-Framework/test-suites/relayboot/report.html"]=file
- ["test-results/Robot-Framework/test-suites/relay-turnoff/log.html"]=file
- ["test-results/Robot-Framework/test-suites/relay-turnoff/output.xml"]=file
- ["test-results/Robot-Framework/test-suites/relay-turnoff/report.html"]=file

**Works fully only with Nvidia targets, modify to work also with x86 targets**
- check_dynamic_files() searches for "*.img.zst" "*.img.zst.sig" patterns only in sd-image sudirectory
- This works for Nvidia targets but not for x86 targets because x86 targets have their "*.img.zst" "*.img.zst.sig" files in the root directory, not in the sd-image subdirectory

karim20230 commented:
> **No need to check these files, remove:**
> 
>     * ["nix-support/hydra-build-products"]=file
> 
>     * ["nix-support/system"]=file
> 
>     * ["esp.offset"]=file
> 
>     * ["esp.size"]=file
> 
>     * ["root.offset"]=file
> 
>     * ["root.size"]=file
> 
> 
> **Boot and turnoff test result directories have changed, these are the new directories, modify:**
> 
>     * ["test-results/Robot-Framework/test-suites/relayboot/log.html"]=file
> 
>     * ["test-results/Robot-Framework/test-suites/relayboot/output.xml"]=file
> 
>     * ["test-results/Robot-Framework/test-suites/relayboot/report.html"]=file
> 
>     * ["test-results/Robot-Framework/test-suites/relay-turnoff/log.html"]=file
> 
>     * ["test-results/Robot-Framework/test-suites/relay-turnoff/output.xml"]=file
> 
>     * ["test-results/Robot-Framework/test-suites/relay-turnoff/report.html"]=file
> 
> 
> **Works fully only with Nvidia targets, modify to work also with x86 targets**
> 
>     * check_dynamic_files() searches for "_.img.zst" "_.img.zst.sig" patterns only in sd-image sudirectory
> 
>     * This works for Nvidia targets but not for x86 targets because x86 targets have their "_.img.zst" "_.img.zst.sig" files in the root directory, not in the sd-image subdirectory

Thanks @ktusawrk! I implemented as per your feedback.

-Removed unnecessary F checks
-Updated test result directories
-Added Support checks for both Nvidia and x86 targets


Here's the result for x86

```
bash ./qa-script.sh https://ghaf-jenkins-controller-release.northeurope.cloudapp.azure.com/artifacts/ghaf-release-pipeline/build_2-commit_45a676abc3e3f028cafd1a78cf4ac7e3104a748f/x86_64-linux.generic-x86_64-debug/
Checking artifacts in https://ghaf-jenkins-controller-release.northeurope.cloudapp.azure.com/artifacts/ghaf-release-pipeline/build_2-commit_45a676abc3e3f028cafd1a78cf4ac7e3104a748f/x86_64-linux.generic-x86_64-debug/...
Checking test-results/Robot-Framework/test-suites/relay-turnoff/output.xml... ✓ Found
Checking test-results/Robot-Framework/test-suites/bat/report.html... ✓ Found
Checking scs/sbom.cdx.json... ✓ Found
Checking test-results/Robot-Framework/test-suites/bat/log.html... ✓ Found
Checking scs/provenance.json... ✓ Found
Checking test-results/Robot-Framework/test-suites/bat/output.xml... ✓ Found
Checking scs/sbom.spdx.json... ✓ Found
Checking test-results/Robot-Framework/test-suites/relay-turnoff/log.html... ✓ Found
Checking scs/sbom.csv... ✓ Found
Checking test-results/Robot-Framework/test-suites/relayboot/output.xml... ✓ Found
Checking test-results/Robot-Framework/test-suites/relayboot/report.html... ✓ Found
Checking scs/provenance.json.sig... ✓ Found
Checking scs/vulns.txt... ✓ Found
Checking test-results/Robot-Framework/test-suites/relay-turnoff/report.html... ✓ Found
Checking test-results/Robot-Framework/test-suites/relayboot/log.html... ✓ Found
Checking *.img.zst... ✓ Found
Checking *.img.zst.sig... ✓ Found
```

ktusawrk reviewed (changes requested):
It appears that generic Lenovo X1 Carbon image file and signature file are named differently than generic x86 image and signature files

**Image files**
- x86 generic: nixos.img.zst
- X1 Carbon:  disk1.raw.zst

**Signature files**
- x86 generic: nixos.img.zst.sig
- X1 Carbon:  disk1.raw.zst.sig

This script does not find X1 Carbon image and signature files
```
Checking *.img.zst... ✗ Missing
Checking *.img.zst.sig... ✗ Missing
```

Add support also for X1 Carbon image and signature files

karim20230 commented:
@ktusawrk  Should we also add support for ` x86_64-linux.lenovo-x1-carbon-gen11-debug-installer`?

Also, I noticed that the artifact image file and the signature file are named differently, and their directories are also different!

ktusawrk commented:
No need to support x86_64-linux.lenovo-x1-carbon-gen11-debug-installer. It is special case. For example, it is not yet automatically tested so there are no automatic test results.
